common/sgx/internal.edl

```diff
@@ -11,17 +11,126 @@
+            size_t qe_id_info_size,
+            [out, count=1] size_t* qe_id_info_size_out,
+            [out, size=issuer_chain_size] void* issuer_chain,
+            size_t issuer_chain_size,
+            [out, count=1] size_t* issuer_chain_size_out);
+
+        uint32_t oe_get_revocation_info_ocall(
+            [in] uint8_t fmspc[6],
+            size_t num_crl_urls,
+            [in, string] const char* crl_urls0,
+            [in, string] const char* crl_urls1,
+            [in, string] const char* crl_urls2,
+            [out, size=tcb_info_size] void* tcb_info,
+            size_t tcb_info_size,
+            [out, count=1] size_t* tcb_info_size_out,
+            [out, size=tcb_issuer_chain_size] void* tcb_issuer_chain,
+            size_t tcb_issuer_chain_size,
+            [out, count=1] size_t* tcb_issuer_chain_size_out,
+            [out, size=crl0_size] void* crl0,
+            size_t crl0_size,
+            [out, count=1] size_t* crl0_size_out,
+            [out, size=crl1_size] void* crl1,
+            size_t crl1_size,
+            [out, count=1] size_t* crl1_size_out,
+            [out, size=crl2_size] void* crl2,
+            size_t crl2_size,
+            [out, count=1] size_t* crl2_size_out,
+            [out, size=crl_issuer_chain0_size] void* crl_issuer_chain0,
+            size_t crl_issuer_chain0_size,
+            [out, count=1] size_t* crl_issuer_chain0_size_out,
+            [out, size=crl_issuer_chain1_size] void* crl_issuer_chain1,
+            size_t crl_issuer_chain1_size,
+            [out, count=1] size_t* crl_issuer_chain1_size_out,
+            [out, size=crl_issuer_chain2_size] void* crl_issuer_chain2,
+            size_t crl_issuer_chain2_size,
+            [out, count=1] size_t* crl_issuer_chain2_size_out);
+
+        void oe_log_ocall(
+            uint32_t log_level,
+            [in, string] const char* message);
+
+        // Translate the addresses in buffer[] into symbol names.
+        // (similar to Linux backtrace_symbols().
+        uint32_t oe_backtrace_symbols_ocall(
+            [user_check] oe_enclave_t* oe_enclave,
+            [in, count=size] const uint64_t* buffer,
+            size_t size,
+            [out, size=strings_buf_size] void* strings_buf,
+            size_t strings_buf_size,
+            [out, count=1] size_t* strings_buf_size_out);
+
+        void oe_thread_wake_wait_ocall(
+            [user_check] oe_enclave_t* oe_enclave,
+            uint64_t waiter_tcs,
+            uint64_t self_tcs);
+
+        void* oe_realloc_ocall(
+            [user_check] void* ptr,
+            size_t size);
+
+        // Write a string to the console (device: (0=stdout, 1=stderr).
+        void oe_write_ocall(
+            int device,
+            [in, string] const char* str,
+            size_t len);
 
-        /* This function returns its value parameter. */
-        int oe_internal_ping_ocall(int value);
+        uint32_t oe_get_cpuid_table_ocall(
```

Comment:
I like this design of enclave pulling the CPUID table.

Reply:
Thanks! It worked around a chicken and egg problem. We cannot use edger8r function tables until we reach a certain level of initialization (which is why the init function is split into one ECALL and one OCALL).